agent/provider/src/cli/keystore.rs

```diff
@@ -131,44 +152,116 @@ fn remove(config: ProviderConfig, remove: Remove) -> anyhow::Result<()> {
 }
 
 fn print_cert_list(config: &ProviderConfig, certs_data: Vec<Cert>) -> anyhow::Result<()> {
-    let mut table = CertTable::new();
+    let mut table_builder = CertTableBuilder::new();
     for data in certs_data {
-        table.add(data);
+        table_builder.add(data);
     }
-    table.print(config)?;
+
+    table_builder.build().print(config)?;
     Ok(())
 }
 
-struct CertTable {
-    table: ResponseTable,
+fn find_ids_by_prefix(certs: &[Cert], prefix: &str) -> Vec<String> {
+    certs
+        .iter()
+        .map(|cert| cert.id())
+        .filter(|id| id.starts_with(prefix))
+        .collect()
 }
 
-impl CertTable {
+struct CertTableBuilder {
+    entries: Vec<Cert>,
+}
+
+impl CertTableBuilder {
     pub fn new() -> Self {
+        Self {
+            entries: Vec::new(),
+        }
+    }
+
+    pub fn add(&mut self, cert: Cert) {
+        self.entries.push(cert)
+    }
+
+    pub fn build(self) -> CertTable {
+        const DIGEST_PREFIX_LENGTHS: [usize; 3] = [8, 32, 128];
+
+        // hard-code support for the use of the entire signature, regardless of its size,
+        // ensure all prefixes are no longer than the signature, and remove duplicates.
+        //
+        // these are, by construction, sorted smallest to largest.
+        let prefix_lengths = |id_len| {
+            DIGEST_PREFIX_LENGTHS
+                .iter()
+                .map(move |&n| std::cmp::min(n, id_len))
+                .chain(std::iter::once(id_len))
+                .dedup()
+        };
+
+        let mut prefix_uses = HashMap::<String, u32>::new();
+        for cert in &self.entries {
+            for len in prefix_lengths(cert.id().len()) {
+                let mut prefix = cert.id();
+                prefix.truncate(len);
+
+                *prefix_uses.entry(prefix).or_default() += 1;
+            }
+        }
+
+        let mut ids = Vec::new();
+        for cert in &self.entries {
+            for len in prefix_lengths(cert.id().len()) {
+                let mut prefix = cert.id();
+                prefix.truncate(len);
+
+                let usages = *prefix_uses
+                    .get(&prefix)
+                    .expect("Internal error, unexpected prefix");
+
+                // the longest prefix (i.e. the entire fingerprint) will be unique, so
+                // this condition is guaranteed to execute during the last iteration,
+                // at the latest.
+                if usages == 1 {
+                    ids.push(prefix);
+                    break;
+                }
+            }
+        }
+
+        let mut values = Vec::new();
+        for (id_prefix, cert) in ids.into_iter().zip(self.entries.into_iter()) {
+            values.push(match cert {
+                Cert::X509(cert) => {
+                    serde_json::json! { [ id_prefix, cert.not_after, cert.subject, cert.permissions ] }
```

Comment:
Output of this table will probably change, but it is out of scope of this task.
I could not find relevant task. Maybe it was not clarified yet.